### PR TITLE
MAINT: style fixes to WaveletTransform

### PR DESCRIPTION
**Pull request by adler-j**
Done so far:

* Change boundary conditions to follow ODL style.
* Made the distinction between ODL and pywt more clear in general.
* Updated tests according to the above.
* Add PyWavelets to intersphinx.

TODO:

Add the actual example.

### CODE REVIEW

**Comment by adler-j** [comment]
Added very simple example. Are we happy with this for now, or should we add discussion on different wavelets etc?

File: examples/trafos/wavelet_trafo.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Simple example on the usage of the Wavelet Transform."""
```
Comment (kohr-h) [inline examples/trafos/wavelet_trafo.py]: The doc should mention that the `PyWavelets` package is needed to run it.

File: odl/trafos/wavelet.py [hunk 2]
```diff
@@ -41,6 +41,60 @@
...
+    elif pad_mode == 'order1':
+        return 'sp1'
+    elif pad_mode == 'symmetric':
+        return 'sym'
+    elif pad_mode == 'periodic':
+        return 'ppd'
+    elif pad_mode == 'pywt_per':
+        return 'per'
+    else:
+        raise ValueError("unknown pad_mode ('{}') and pad_const ({})"
```
Comment (kohr-h) [inline odl/trafos/wavelet.py]: no parentheses
Probably a check for `'constant'` pad mode and constant != 0 would give a better error message.

File: odl/trafos/wavelet.py [hunk 3]
```diff
@@ -56,13 +110,13 @@ def coeff_size_list(shape, nscales, wbasis, pad_mode):
...
         transform.  This parameter is checked against the maximum number of
-        scales returned by ``pywt.dwt_max_level``. For more information
+        scales returned by `pywt.dwt_max_level`. For more information
         see the `PyWavelets documentation on the maximum level of scales
         <http://www.pybytes.com/pywavelets/ref/\
 dwt-discrete-wavelet-transform.html#maximum-decomposition-level\
 -dwt-max-level>`_.
 
-    wbasis : ``pywt.Wavelet``
+    wbasis : {string, `pywt.Wavelet`}
```
Comment (kohr-h) [inline odl/trafos/wavelet.py]: Curly brackets are used for a fixed range of *values*, not *types*.
Can we link to `pywt.Wavelet`?

File: odl/trafos/wavelet.py [hunk 4]
```diff
@@ -525,71 +585,82 @@ def __init__(self, domain, nscales, wbasis, pad_mode):
...
             Describes properties of a selected wavelet basis.
             See PyWavelet `documentation
             <http://www.pybytes.com/pywavelets/ref/wavelets.html>`_
 
             Possible wavelet families are:
 
-            Haar (``haar``)
+            'Haar': Haar
+
+            'db'  : Daubechies
```
Comment (kohr-h) [inline odl/trafos/wavelet.py]: Help pages have a non-proportional font typically, so the extra spaces look wrong there.

File: odl/trafos/wavelet.py [hunk 5]
```diff
@@ -525,71 +585,82 @@ def __init__(self, domain, nscales, wbasis, pad_mode):
...
 
-            Biorthogonal (``bior``)
+            'rbio': Reverse biorthogonal
 
-            Reverse biorthogonal (``rbio``)
+            'dmey': Discrete FIR approximation of Meyer wavelet
 
-            Discrete FIR approximation of Meyer wavelet (``dmey``)
+        pad_mode : {'constant', 'symmetric', 'periodic', 'order0', 'order1',
+                    'pywt_periodization'}
```
Comment (kohr-h) [inline odl/trafos/wavelet.py]: different from the helper above (`pywt_per` there)

File: odl/trafos/wavelet.py [hunk 6]
```diff
@@ -525,71 +585,82 @@ def __init__(self, domain, nscales, wbasis, pad_mode):
...
+            ``'order0'``: Extend constantly with the outmost values
+            (ensures continuity).
 
-            'sym': symmetric padding -- signal extension by mirroring samples
+            ``'order1'``: Extend with constant slope (ensures continuity of
+            the first derivative). This requires at least 2 values along
+            each axis where padding is applied.
 
-            'ppd': periodic padding -- signal is trated as a periodic one
+            ``'pywt_per'``:  like ``'periodic'``-padding but gives the smallest
```
Comment (kohr-h) [inline odl/trafos/wavelet.py]: what gives?
e.g. "this variant"

File: odl/trafos/wavelet.py [hunk 7]
```diff
@@ -525,71 +585,82 @@ def __init__(self, domain, nscales, wbasis, pad_mode):
...
+        self.wbasis = wbasis
+        self.pywt_wbasis = pywt_wbasis(wbasis)
 
-        if isinstance(wbasis, pywt.Wavelet):
-            self.wbasis = wbasis
-        else:
-            self.wbasis = pywt.Wavelet(wbasis)
+        self.impl = str(impl).lower()
+        if self.impl != 'pywt':
+            raise ValueError('impl not understood')
```
Comment (kohr-h) [inline odl/trafos/wavelet.py]:
```
`impl`
```

**Comment by kohr-h** [comment]
Very good changes, I'll run the example when I'm done with some other stuff. In general, I don't know if we need anything more complicated that this. In the long run, some L1-on-wavelet-coefficents type example would be nice.

**Comment by adler-j** [comment]
So, do we move the example code out of this and close this guy in favor of the much larger one #547?

**Comment by kohr-h** [comment]
Yes, I'll cherry-pick over the commit.
